java/vector/src/main/java/org/apache/arrow/vector/types/pojo/Field.java

```diff
@@ -177,18 +177,21 @@ public TypeLayout getTypeLayout() {
   }
 
   @Override
+  public int hashCode() {
+    return Objects.hash(name, nullable, type, dictionary, children);
+  }
+
+  @Override
   public boolean equals(Object obj) {
     if (!(obj instanceof Field)) {
       return false;
     }
     Field that = (Field) obj;
     return Objects.equals(this.name, that.name) &&
-            Objects.equals(this.nullable, that.nullable) &&
-            Objects.equals(this.type, that.type) &&
+           Objects.equals(this.nullable, that.nullable) &&
+           Objects.equals(this.type, that.type) &&
            Objects.equals(this.dictionary, that.dictionary) &&
-            (Objects.equals(this.children, that.children) ||
-                    (this.children == null || this.children.size() == 0) &&
-                    (that.children == null || that.children.size() == 0));
```

Comment:
Since the constructor always assigns an empty list when children == null this would never happen.
children is never null.